### PR TITLE
Update README.md

### PR DESCRIPTION
**Pull request by Ngoguey42**
It's review time for the new README, the new repository's description and the new tags. The documentation will be merge next

https://github.com/airware/buzzard/blob/README-update/README.md

### CODE REVIEW

**Comment by Ngoguey42** [comment]
@gaslen, @j-bd can you review too?

File: README.md [hunk 1]
```diff
@@ -1,103 +1,216 @@
...
+```
+
+
+```
+# repository's keywords
+X python gis raster vector gdal feedback-welcome footprint
+? python gis geospatial raster image vector geometry gdal ogr osr data-science footprint rasters-pipelines
+
+more? less?
+```
```
[the remark this answers is not in the document]
Comment (Ngoguey42) [reply README.md]: Googling `footprint raster` gives non-buzzard results related to GIS

File: README.md [hunk 2]
```diff
@@ -1,103 +1,216 @@
...
+
```
[the remark this answers is not in the document]
Comment (Ngoguey42) [reply README.md]: what about `rasters-pipelines`?

File: README.md [hunk 3]
```diff
@@ -1,103 +1,216 @@
...
+Begining of `README.md`.
+
+---
+
 # `buzzard`
+:heavy_multiplication_x: In a nutshell, `buzzard` reads and writes geospatial raster and vector data.
 
-In a nutshell, `buzzard` reads and writes geospatial raster and vector data.
+:grey_question: In a nutshell, `buzzard` provides powerful abstractions to work with images and geometries that comes from many kind of sources.
```
[the remark this answers is not in the document]
Comment (Ngoguey42) [reply README.md]: nor does `raster`!

File: README.md [hunk 4]
```diff
@@ -1,103 +1,216 @@
...
+:heavy_multiplication_x: :heavy_multiplication_x: :heavy_multiplication_x:
+
+:grey_question: :grey_question: :grey_question:
+- A _python_ library.
+- Primarily designed to hide all cumbersome operations when doing data-science with [GIS](https://en.wikipedia.org/wiki/Geographic_information_system) files.
+- Multipurpose, it can be used in all kind of situations were images or geometries are involved.
+- A pythonic wrapper for _osgeo_'s _gdal_/_ogr_/_osr_.
+- A solution to work with arbitrary large images by simplifying and automating the manipulation of image slices.
+- Developed at [Delair](https://delair.aero) where it is used in several deep learning and algorithmic projects.
+
```
[the remark this answers is not in the document]
Comment (Ngoguey42) [reply README.md]: What about `A pythonic wrapper for _osgeo_'s _gdal_/_ogr_/_osr_.`?

File: README.md [hunk 5]
```diff
@@ -1,103 +1,216 @@
...
+:grey_question: :grey_question: :grey_question:
 
-## Simple example
-This example illustrates visualization of a raster based on polygons.
+## How to open and read files
+This example demonstrates how to visualize a large raster based on polygons.
```
[the remark this answers is not in the document]
Comment (Ngoguey42) [reply README.md]: What about `polygon per polygon`?

File: README.md [hunk 6]
```diff
@@ -1,103 +1,216 @@
...
+km = keras.models.load_model('path/to/deep-learning-model.hdf5')
+
+# Chunk the raster's Footprint to Footprints of size
+# 1920 x 1080 pixel stored in a 2d numpy array
+tiles = r.fp.tile(1920, 1080)
+
+all_roads = []
+
+for i, fp in enumerate(tiles.flat):
+    rgb = r.get_data(fp=fp, channels=(0, 1, 2))
```
[the remark this answers is not in the document]
Comment (Ngoguey42) [reply README.md]: :-1: !!!

[the remark this answers is not in the document]
Comment (Ngoguey42) [reply README.md]: It changes, and there is an alpha channel here

File: README.md [hunk 7]
```diff
@@ -1,103 +1,216 @@
...
+## `buzzard` contains
+:heavy_multiplication_x: :heavy_multiplication_x: :heavy_multiplication_x:
+- a class to open/read/write/create GIS files: [`Dataset`](./buzzard/_dataset.py)
+- a toolbox class designed to locate a rectangle in both image space and geometry space: [`Footprint`](./buzzard/_footprint.py)
+
+:heavy_multiplication_x: :heavy_multiplication_x: :heavy_multiplication_x:
+
+:grey_question: :grey_question: :grey_question:
+- A [`Dataset`](https://github.com/airware/buzzard/blob/master/buzzard/_dataset.py) class that oversees all opened files in order to share ressources.
+- An immutable toolbox class, the [`Footprint`](https://github.com/airware/buzzard/blob/master/buzzard/_footprint.py), designed to locate a rectangle in both image space and geometry space.
```
[the remark this answers is not in the document]
Comment (Ngoguey42) [reply README.md]: You're right, the Footprint objects are immutable, not the class itself. No one will notice!

File: README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (Toz3wm) [reply README.md]: meh.. Ô wise @azarz, enlight us !

File: README.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (Toz3wm) [reply README.md]: :+1:

File: README.md [hunk 8]
```diff
@@ -1,103 +1,216 @@
...
+# 1920 x 1080 pixel stored in a 2d numpy array
+tiles = r.fp.tile(1920, 1080)
+
+all_roads = []
+
+for i, fp in enumerate(tiles.flat):
+    rgb = r.get_data(fp=fp, channels=(0, 1, 2))
+
+    # Perform pixelwise semantic segmentation with a keras model
+    predictions_heatmap = km.predict(rgb[np.newaxis, ...])[0]
```
[the remark this answers is not in the document]
Comment (Ngoguey42) [reply README.md]: Have you heard about aerial intelligence?

File: README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (azarz) [reply README.md]: Thou summoneth me, hither I cometh.
When searching `raster pipelines` on the mighty Google, there is one image result that seems to match what buzzard does.
Also, to answer Toz3wm's initial comment, I don't think putting "how" the work is done (optimizations) in the tags/keywords is necessary.

File: README.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (Toz3wm) [reply README.md]: We should fund an agency on that topic

File: README.md [hunk 9]
```diff
@@ -1,103 +1,219 @@
...
+
+Begining of `README.md`.
+
+---
+
 # `buzzard`
+:heavy_multiplication_x: In a nutshell, `buzzard` reads and writes geospatial raster and vector data.
 
-In a nutshell, `buzzard` reads and writes geospatial raster and vector data.
+:grey_question: In a nutshell, `buzzard` provides powerful abstractions to work with images and geometries that comes from many kind of sources.
```
Comment (gaslen) [inline README.md]: Do you really want to remove `geospatial` ?

File: README.md [hunk 10]
```diff
@@ -1,103 +1,219 @@
...
 ```py
 import buzzard as buzz
 import numpy as np
 import matplotlib.pyplot as plt
 
-rgb_path = 'path/to/raster.file'
-polygons_path = 'path/to/vector.file'
+# Open the files. Only files' metadata are read so far
+r = buzz.open_raster('path/to/rgba-image.file')
+v = buzz.open_vector('path/to/polygons.file')
```
Comment (gaslen) [inline README.md]: Not a fan of single character for variables names, especially in a README

File: README.md [hunk 11]
```diff
@@ -1,103 +1,219 @@
...
     rgb[mask] = (rgb[mask] * 0.5).astype(np.uint8)
-    rgb[alpha == 0] = [255, 0, 0]
+    rgb[alpha == 0] = [236, 120, 57]
 
+    # Show the result with matplotlib
     plt.imshow(rgb)
     plt.show()
+
 ```
+Images from the ISPRS's Potsdam dataset.
```
Comment (gaslen) [inline README.md]: Make an hyperlink ?
http://www2.isprs.org/commissions/comm3/wg4/2d-sem-label-potsdam.html

File: README.md [hunk 12]
```diff
@@ -1,103 +1,219 @@
...
+:heavy_multiplication_x: :heavy_multiplication_x: :heavy_multiplication_x:
 - a python library
 - a `gdal`/`ogr`/`osr` wrapper
 - designed to hide all cumbersome operations while working with GIS files
 - designed for data science workflows
 - under active development (see [`todo`](https://www.notion.so/buzzard/2c94ef6ee8da4d6280834129cc00f4d2?v=334ead18796342feb32ba85ccdfcf69f))
 - tested with `pytest` with python 3.4 and python 3.7
 
-## `buzzard` contains
-- a class to open/read/write/create GIS files: [`Dataset`](./buzzard/_dataset.py)
```
Comment (j-bd) [inline README.md]: From my point of view, it would have been nice to keep a short description of the main classes for Buzzard beginners. Indeed, geospatial field is easily confusing. A technical word may bring to different interpretation.

Furthermore, we now jump straight from a general presentation to examples. Developing the explanation of classes could make easier to get the importance of buzzard even for GIS specialist.

Comment (j-bd) [reply README.md]: I have just noticed you moved this part at the end. I still believe we should keep it before the examples and bring more explanation ...

File: README.md (hunk 9 above)
Comment (Ngoguey42) [reply README.md]: I do. The `buzzard is` section below is quite honest already, but I would like the description and this intro to also reflect what buzzard may become

File: README.md (hunk 10 above)
Comment (Ngoguey42) [reply README.md]: I'm not sure about this one. It makes the buzzard syntax look lighter.
And in this context, it is easy to infer the `r` and `v` meanings because of the perfect horizontal symmetry with the `raster` and `vector` words 14 chars to the right. 

What do you think?

File: README.md (hunk 12 above)
Comment (Ngoguey42) [reply README.md]: Good idea

**Comment by Ngoguey42** [comment]
@sophiefnt Can you review too please? :smile:

File: README.md [hunk 13]
```diff
@@ -1,103 +1,220 @@
...
-- More tools, syntaxes and algorithms to work with raster datasets that don't fit in memory
-- Strong support of non north-up / west-left footprints
-- Data visualization tools
-- Strong performance improvements
-- Floating point precision loss handling improvements
+:heavy_multiplication_x: :heavy_multiplication_x: :heavy_multiplication_x:
+
+:grey_question: :grey_question: :grey_question:
+- Raster and vector files opening and creation. Supports all [GDAL drivers (GTiff, PNG, ...)](https://www.gdal.org/formats_list.html) and all [OGR drivers (GeoJSON, DXF, Shapefile, ...)](https://www.gdal.org/ogr_formats.html).
+- Raster files reading to _numpy.ndarray_.
```
[the remark this answers is not in the document]
Comment (sophiefnt) [reply README.md]: clarification? don't you read the raster ""from"" a numpy array ? and then (cf. line below) you write the raster file as a numpy array (i.e. write the raster in that format)?

File: README.md [hunk 14]
```diff
@@ -1,103 +1,220 @@
...
+  - _Options:_ `sub-rectangle reading`, `rotated and scaled sub-rectangle reading (thanks to on-the-fly remapping with OpenCV)`, `automatic parallelization of read and remapping (soon)`, `async (soon)`, `be the source of an image processing pipeline (soon)`.
+  - _Properties:_ `thread-safe`
+- Raster files writing from _numpy.ndarray_.
+  - _Options:_ `sub-rectangle writing`, `rotated and scaled sub-rectangle writing (thanks to on-the-fly remapping with OpenCV)`, `masked writing`.
+- Vector files reading to _shapely objects_, _geojson dict_ and _raw coordinates_.
```
[the remark this answers is not in the document]
Comment (sophiefnt) [reply README.md]: same comment as above: don't you read the files "" from"" these objects and then you write (new) vector files to these objects ? 
I'm a little confused

File: README.md [hunk 15]
```diff
@@ -1,103 +1,220 @@
...
+  - _Options:_ `masking`.
+  - _Properties:_ `thread-safe`
+- Vector files writing from _shapely objects_, _geojson dict_ and _raw coordinates_.
+- Powerful manipulations of raster windows
+- Instantiation of image processing pipelines where each node is a raster, and each edge is a user defined python function working on _numpy.ndarray_ (beta, partially implemented).
```
[the remark this answers is not in the document]
Comment (Ngoguey42) [reply README.md]: The [instantiation](https://en.wiktionary.org/wiki/instantiation) concern the [pipelines](https://en.wikipedia.org/wiki/Pipeline_(computing)), not the images. Here `instantiation` refer to the act of creating an object that can ingest images and spit out new images. 

For example a deep learning model for computer vision is an `image processing pipeline`, you first instanciate it, and then give it an image that will be transformed step by step. 

In buzzard it means chaining the calls to `create_cached_raster_recipe`.

@sophiefnt What can be changed to make it clearer? Maybe add a link to the `Dataset.create_cached_raster_recipe` method?

File: README.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (Ngoguey42) [reply README.md]: I edited your above suggestion, is that clearer?

File: README.md [hunk 16]
```diff
@@ -1,103 +1,220 @@
...
-- Strong performance improvements
-- Floating point precision loss handling improvements
+:heavy_multiplication_x: :heavy_multiplication_x: :heavy_multiplication_x:
+
+:grey_question: :grey_question: :grey_question:
+- Raster and vector files opening and creation. Supports all [GDAL drivers (GTiff, PNG, ...)](https://www.gdal.org/formats_list.html) and all [OGR drivers (GeoJSON, DXF, Shapefile, ...)](https://www.gdal.org/ogr_formats.html).
+- Raster files reading to _numpy.ndarray_.
+  - _Options:_ `sub-rectangle reading`, `rotated and scaled sub-rectangle reading (thanks to on-the-fly remapping with OpenCV)`, `automatic parallelization of read and remapping (soon)`, `async (soon)`, `be the source of an image processing pipeline (soon)`.
+  - _Properties:_ `thread-safe`
+- Raster files writing from _numpy.ndarray_.
```
[the remark this answers is not in the document]
Comment (Ngoguey42) [reply README.md]: I edited your above suggestion, is that clearer?

File: README.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (Ngoguey42) [reply README.md]: I edited your above suggestion, is that clearer?

File: README.md [hunk 17]
```diff
@@ -1,103 +1,220 @@
...
+  - _Options:_ `sub-rectangle writing`, `rotated and scaled sub-rectangle writing (thanks to on-the-fly remapping with OpenCV)`, `masked writing`.
+- Vector files reading to _shapely objects_, _geojson dict_ and _raw coordinates_.
+  - _Options:_ `masking`.
+  - _Properties:_ `thread-safe`
+- Vector files writing from _shapely objects_, _geojson dict_ and _raw coordinates_.
```
[the remark this answers is not in the document]
Comment (Ngoguey42) [reply README.md]: I edited your above suggestion, is that clearer?

File: README.md [hunk 18]
```diff
@@ -1,103 +1,220 @@
...
+:heavy_multiplication_x: :heavy_multiplication_x: :heavy_multiplication_x:
 - a class to open/read/write/create GIS files: [`Dataset`](./buzzard/_dataset.py)
 - a toolbox class designed to locate a rectangle in both image space and geometry space: [`Footprint`](./buzzard/_footprint.py)
 
-## Simple example
-This example illustrates visualization of a raster based on polygons.
+:heavy_multiplication_x: :heavy_multiplication_x: :heavy_multiplication_x:
+
+:grey_question: :grey_question: :grey_question:
+- A [`Dataset`](https://github.com/airware/buzzard/blob/master/buzzard/_dataset.py) class that oversees all opened files in order to share ressources.
```
[the remark this answers is not in the document]
Comment (Ngoguey42) [reply README.md]:
```suggestion
- A [`Dataset`](https://github.com/airware/buzzard/blob/master/buzzard/_dataset.py) class that oversees all opened raster and vector files in order to share resources.
```

File: README.md [hunk 19]
```diff
@@ -1,103 +1,220 @@
...
+
+Begining of `README.md`.
+
+---
+
 # `buzzard`
+:heavy_multiplication_x: In a nutshell, `buzzard` reads and writes geospatial raster and vector data.
 
-In a nutshell, `buzzard` reads and writes geospatial raster and vector data.
+:grey_question: In a nutshell, `buzzard` provides powerful abstractions to work with images and geometries that comes from many kind of sources.
```
[the remark this answers is not in the document]
Comment (Ngoguey42) [reply README.md]: Smart!

[the remark this answers is not in the document]
Comment (Ngoguey42) [reply README.md]:
```suggestion
:grey_question: In a nutshell, the `buzzard` library provides powerful abstractions to manipulate together images (GeoTIFF, PNG, ...) and geometries (GeoJSON, Shapefile, ...) that come from different kind of sources.
```

File: README.md (hunk 18 above)
[the remark this answers is not in the document]
Comment (Ngoguey42) [reply README.md]: I was referring to this: https://github.com/airware/buzzard/blob/master/buzzard/_dataset.py#L33

What if I keep it short with just `- A [`Dataset`](https://github.com/airware/buzzard/blob/master/buzzard/_dataset.py) class that oversees all opened raster and vector files.` ?

File: README.md (hunk 15 above)
[the remark this answers is not in the document]
Comment (sophiefnt) [reply README.md]: ok get it, thanks ! So, the nodes and edges are later fed to the pipelines? we just create the different steps here but don't transform the image yet ? 
I think that's ok like that, if the vocab is widely-used

File: README.md (hunk 19 above)
[the remark this answers is not in the document]
Comment (Ngoguey42) [reply README.md]: now: 
- In a nutshell, the `buzzard` library provides powerful abstractions to manipulate together images (GeoTIFF, PNG, ...) and geometries (GeoJSON, Shapefile, ...) that come from different kind of sources

I feel that the `that come from different kind of sources` part is a bit redundant.
What about: 
- In a nutshell, the `buzzard` library provides powerful abstractions to manipulate together images and geometries that come from different kind of sources (`GeoTIFF`, `PNG`, `GeoJSON`, `Shapefile`, `numpy array`, `buzzard pipelines`, ...).

@azarz @sophiefnt ?

[the remark this answers is not in the document]
Comment (sophiefnt) [reply README.md]: yes, agree, and it doesn't break the sentence so it's easier to read !
(except if you meant by "sources" where the actual images come from, i.e. drone acquired ?  )

[the remark this answers is not in the document]
Comment (Ngoguey42) [reply README.md]: No, by `source` I mean what is contained between the parenthesis
